### ISSUE TITLE
Text scrolling unexpectedly when highlighting from corner

### ISSUE THREAD

**Issue by jonnyl2** [koreader/koreader#11223]
* KOReader version: 2023.10-52
* Device: Kobo Aura H2O

#### Issue

Very often, when starting a highlight in the upper left hand corner of the screen, the text scrolls backward even though I'm moving in the forward direction and do not intend to highlight anything from the previous page. This interferes with my highlighting process. It doesn't happen every time, but often enough that it's rather bothersome. (Increasing the margin would prevent it but  that would waste valuable screen space.)

https://github.com/koreader/koreader/assets/95502269/60fa47d4-540c-4895-bac3-db4aed0abbfa

#### Steps to reproduce

1. Start highlighting from upper left corner in the forward direction
2. Observe text moving backwards

#### Suggested solution

If there is no easy fix without unintended side effects (or even if there is), please consider making the text-scrolling-while-highlighting feature an option that can be disabled entirely. In my opinion, the `Select` feature is much better suited for inter-page highlights on e-readers and makes the scrolling unnecessary.

##### `crash.log` (if applicable)

Will gladly provide if needed.

Thank you for your help.

**Comment by poire-z** [comment]
We could require the user to have been "not in a corner" before "going in a corner" to allow this feature to work.

Quick patch, how would that work for you?
```diff
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -1181,6 +1181,8 @@ function ReaderHighlight:onHold(arg, ges)
     end
     self.gest_pos = self.hold_pos

+    self.allow_hold_pan_corner_scroll = false -- reset this, don't allow that yet
+
     -- check if we were holding on an image
     -- we provide want_frames=true, so we get a list of images for
     -- animated GIFs (supported by ImageViewer)
@@ -1300,7 +1302,15 @@ function ReaderHighlight:onHoldPan(_, ges)
             is_in_next_page_corner = self.holdpan_pos.y > 7/8*self.screen_h
                                       and self.holdpan_pos.x > 7/8*self.screen_w
         end
-        if is_in_prev_page_corner or is_in_next_page_corner then
+        if not self.allow_hold_pan_corner_scroll then
+            if not is_in_prev_page_corner and not is_in_next_page_corner then
+                -- We expect the user to come from a non-corner zone into a corner
+                -- to enable this; this allows normal highlighting without scrolling
+                -- if the selection is started in the corder: the user will have to
+                -- move out from and go back in to trigger a scroll.
+                self.allow_hold_pan_corner_scroll = true
+            end
+        elseif is_in_prev_page_corner or is_in_next_page_corner then
             self:_resetHoldTimer()
             if self.was_in_some_corner then
                 -- Do nothing, wait for the user to move his finger out of that corner
```

It may be a bit bothering for those who want quick, it as (to not duplicate all the corner detection also in onHold) it expects another hold_pan step outside a corner, so if you long press in the middle and go immediately to the corner, it may not trigger. Dunno if it's noticable on devices ? (I may have large intervals/delays on my emulator, not sure.)

May also trigger if you want to do small text selection in a corner, and you go in and out an in that corner.

**Comment by jonnyl2** [comment]
Works perfectly for me, thank you!

>if you long press in the middle and go immediately to the corner, it may not trigger

Didn't observe this behavior at all; it always scrolled as expected.

>May also trigger if you want to do small text selection in a corner, and you go in and out an in that corner.

You could just prevent the scrolling by starting from the corner, or am I missing something?

**Comment by poire-z** [comment]
> You could just prevent the scrolling by starting from the corner, or am I missing something?

You are missing that _you_ don't like/use that autoscroll feature - but people who do may not want it to just not work when started in a corner :)

I don't highlight, but I select sometimes for translation/wikipedia. I'm used for years to the old way of continuous text selection, never really got to use (and possibly like) the 2 steps option. I would not like to have to use one way when selecting in a page, and having to think about using another way when stuff happens in a corner. (And I sometimes have the tedious and comical way of shaking my finger without losing contact with the screen to trigger pan and scroll or stuff...)

I would not mind a menu item to disable corner scroll completely for people who find it annoying - but I think it's fine to have it enabled by default so users see it's possible to do stuff that way (even if it works only with EPUBs and not PDF).

**Comment by jonnyl2** [comment]
>You are missing that you don't like/use that autoscroll feature - but people who do may not want it to just not work when started in a corner :)

No, I understand that. I just didn't understand what you meant here:

>May also trigger if you want to do small text selection in a corner, and you go in and out an in that corner.

Did you mean the use case where the user begins a highlight, say, at the last word on a page and _wants_ it to go onto the next page, but with this patch the scroll wouldn't trigger anymore? That's certainly possible and I didn't think about that. 

I'd also be in favor of an option to disable it completely (ideally with Dispatcher entry, akin to `ignore long-press on corners`). The added advantage compared to this patch would be that I wouldn't have to disable corner long-press when highlighting starting from away from the corner panning towards the corner.

**Comment by poire-z** [comment]
> Did you mean the use case where the user begins a highlight, say, at the last word on a page and wants it to go onto the next page, but with this patch the scroll wouldn't trigger anymore? 

I didn't mean that: this case is fine, one just has to go out of the corner then back (the comical gymnastic I mentionned - it happens often that I have to in-out-in-out-in to get the pan to trigger - there is a minimal distance I think - to get my corner scroll and be able to continue the selection to words on the other page).

I meant when wanting to select words from the start of first line of a page, deciding (as you keep long-pressing) to go to the right, then deciding to go back in the corner to select words from the second line.
![image](https://github.com/koreader/koreader/assets/24273478/c7e80524-4b65-45ae-8d92-ead81910d1ff)
This would trigger out=>in corner and cause a scroll.

**Comment by jonnyl2** [comment]
I see, thanks for explaining. I tested it out and yes, that could certainly be annoying in practice.

**Comment by jonnyl2** [comment]
>I meant when wanting to select words from the start of first line of a page, deciding (as you keep long-pressing) to go to the right, then deciding to go back in the corner to select words from the second line.

@poire-z: I upgraded to the latest nightly so I tried this again without the above patch. And as far as I can tell, your described scenario already happens as it is, so the patch doesn't really change (worsen) that behavior? Or is there something else I'm not seeing ;)

**Comment by poire-z** [comment]
Yes, it already happened. I was just thinking about all the past/present/future little issues we may be having with this corner scroll feature. As we were solving one, I was mentionning the others I've witnessed, so may be our collaborative H.I. could get some ideas :)
(I'm soon off, and won't rush the above patch into the coming up stable release, so it will wait for next year to go in - may be with a menu item to disable the corner scroll feature - haven't really looked in which menu it could go in, may be you'll have some idea?)

**Comment by jonnyl2** [comment]
>may be our collaborative H.I. could get some ideas :)

Sounds good. Going back to your scenario where you'd pan back into the corner while selecting. Wouldn't the only time you'd be doing that be when you've selected the wrong starting word? So that means you'd have to release and repress the finger anyway, and won't encounter the unintended scroll?

>may be with a menu item to disable the corner scroll feature - haven't really looked in which menu it could go in, may be you'll have some idea?

I would put it as the 2nd item under Gear -> Taps and gestures -> Long-press on text ->. (Underneath `☑ Dictionary on single press selection`). 

>I'm soon off

Happy Holidays!

**Comment by poire-z** [comment]
Thanks :)

> I would put it as the 2nd item under Gear -> Taps and gestures -> Long-press on text ->. (Underneath ☑ Dictionary on single press selection).

That would shift 1 of the 9 actions to the next menu page :/
But it could be put at the end of this menu, with the other 2 highlight reated menu items on the 2nd page (I didn't find these when having a quick look :) but it feels right at place with them).

> Going back to your scenario where you'd pan back into the corner while selecting. Wouldn't the only time you'd be doing that be when you've selected the wrong starting word? So that means you'd have to release and repress the finger anyway, and won't encounter the unintended scroll?

Another scenario (which is more my use case) is when selecting for Wikipedia lookup a few words crossing page, say "Romeo and" at bottom right of previous page, and "Juliet" at the top of next page. If selecting "Juliet" in the corner, I need to pan out of that corner and go back in to trigger the corner scroll (I had/have to do that with or without the patch) and be able to select "Romeo and".

**Comment by jonnyl2** [comment]
>Same to (all of) you!

Thank you!

>That would shift 1 of the 9 actions to the next menu page :/
But it could be put at the end of this menu, with the other 2 highlight reated menu items on the 2nd page (I didn't find these when having a quick look :) but it feels right at place with them).

At the end works too, and if those 3 items are together that makes more sense anyway. Regarding the `Highlight very-long-press interval:`, I noticed that it always triggers the 'ask with popup dialog' option. This is fine with me personally as that would be my choice anyway, but shouldn't it rather do whichever action is selected in the menu for a multiple word selection? (Also shouldn't it rather be "Selection very-long-press interval"?)

I was also thinking that it would make for some improved UX if feedback were given when the very-long-press interval has elapsed. For example, why not execute the action immediately when the time has elapsed, rather than wait for the user to release the finger (who has to guess when the time is up)? Or another option would be a quick flash of the selected text, upon which the user knows he can now release the finger.

>Another scenario (which is more my use case) is when selecting for Wikipedia lookup a few words crossing page, say "Romeo and" at bottom right of previous page, and "Juliet" at the top of next page. If selecting "Juliet" in the corner, I need to pan out of that corner and go back in to trigger the corner scroll (I had/have to do that with or without the patch) and be able to select "Romeo and".

Yes, I got that. But would you say it's bothersome to do it that way? I tried and it immediately felt quite intuitive to me to actuate the scrolling that way (and with very little chance it would be done inadvertently). The only other option I could think of is to delay the scrolling by a 2 second delay or something, although I'm not a big fan of imposed delays on the user. (However in this case I could live with it because this situation should occur infrequently enough.)

**Comment by poire-z** [comment]
> Yes, I got that. But would you say it's bothersome to do it that way? I tried and it immediately felt quite intuitive to me 

So it is for me: not bothersome and intuitive, and I want to keep that. So, this behaviour we both like causes the "issue" related above in my post with the 1..2..3 screenshot. Fine by me anyway :)

> Regarding the Highlight very-long-press interval:, I noticed that it always triggers the 'ask with popup dialog' option. This is fine with me personally as that would be my choice anyway, but shouldn't it rather do whichever action is selected in the menu for a multiple word selection?

Well, that's its purpose. The action selected is what happens with "long-press". "very-long-press" is for when you need another action than the "long-press" associated one, so it gives you this popup dialog, so you can select any other rarely needed action.
(If you were confusing "long-press" and "very long-press" intervals, it is indeed a bit clumsy. "very long press interval" is where you found it, a bit related to this highlight feature. "long-press interval" is under `Tap & gestures > Gesture intervals` as it is a bit more generic (long-press is involved in other things than text selection, ie buttons, menu items, ImageViewer...)

> (Also shouldn't it rather be "Selection very-long-press interval"?)

Probably. Our english wording accuracy has not always been what it is today :)

> For example, why not execute the action immediately when the time has elapsed, rather than wait for the user to release the finger (who has to guess when the time is up)?

I'm not sure when the countdown starts, after the initial touch or after pan and the finger is not moving anymore. But who are we to decide that the user is done selecting :) He may be slow, wondering what to include, trying to "fix" its selection.
(And I indeed to count in my head, 1 ... 2 ... 3 ...release.)

> Or another option would be a quick flash of the selected text, upon which the user knows he can now release the finger.

Could be nice. Dunno how complicated it would be to implement, refresh/repaint stuff may be tricky.

**Comment by jonnyl2** [comment]
> Well, that's its purpose. The action selected is what happens with "long-press". "very-long-press" is for when you need another action than the "long-press" associated one, so it gives you this popup dialog, so you can select any other rarely needed action.

I see. I didn't fully grasp its purpose. It never occurred to me that you could use it on larger selections; for me it was just a way to circumvent dictionary pop-ups on single word selections.

>>why not execute the action immediately when the time has elapsed

>But who are we to decide that the user is done selecting :) 

Knowing the above, I take this suggestion back of course :)



>>  (Also shouldn't it rather be "Selection very-long-press interval"?)

> Probably. Our english wording accuracy has not always been what it is today :)

It's not really the English that is inaccurate (a selection _does_ highlight the text after all). It just doesn't correspond with KOReader's feature name (a selection only becomes a highlight after it is added to the bookmark record). But I've also often used "highlight" incorrectly (e.g., see the issue's title), until your responses in this very thread taught me better :)

**Comment by jonnyl2** [comment]
>> Or another option would be a quick flash of the selected text

>Could be nice. Dunno how complicated it would be to implement, refresh/repaint stuff may be tricky.

A whole block of text flashing would also not look very pleasant. Another idea would be to flash that little black rectangle (used for internal links) in the margin on the level of the gesture point when the time has elapsed.

**Comment by poire-z** [comment]
If you want to test it: [highlightTweaks1.zip](https://github.com/koreader/koreader/files/13801561/highlightTweaks1.zip)
Includes:
- the tweak mentionned above at https://github.com/koreader/koreader/issues/11223#issuecomment-1852695604
- a new menu item to allow disabling corner scroll:
![image](https://github.com/koreader/koreader/assets/24273478/dd4730aa-c771-4190-8dce-5ae4ce6425be)
- some icon when very-long-press reached (using the same code than handles the select_mode icon):
![image](https://github.com/koreader/koreader/assets/24273478/9ce6fe34-e2e7-4b02-95eb-d386621a25bf)
(I re-used an existing icon that could feel like a long-press to people with imagination, other ideas - and icon - welcomed. This one is not really eye catching, which may be nice, dunno...)
For now, it is always displayed when very-long-press timeout reached, even if there is no change of behaviour vs. small-long-press. Should it only display when there is a change of behaviour (that is, when `[x] Dictionary on single word selection` is not disabled and select is single word, or when the action is different from "Ask with popup dialog").

**Comment by jonnyl2** [comment]
>some icon when very-long-press reached (using the same code than handles the select_mode icon)

I like it! Obviously I am biased now, but I think its meaning can be easily guessed by any user familiar with the 'very-long press' feature. 

>For now, it is always displayed when very-long-press timeout reached, even if there is no change of behaviour vs. small-long-press. Should it only display when there is a change of behaviour (that is, when [x] Dictionary on single word selection is not disabled and select is single word, or when the action is different from "Ask with popup dialog").

I would leave it the way it is. Adding extra checks/conditions seems unnecessary and provides little (if any) benefit. Users are unlikely to trigger the very-long unless intended (even the minimum 2.5s I set it to seems really long to keep the finger in place unintentionally).

By the way, what would you think about offering the whole range of actions for both long press and very-long press? So the user could choose any one for either gesture. (Personally, I'd rather have the context menu for regular long-press selections and the direct highlight for very-long press.) Not important. Just an idea :)

**Comment by poire-z** [comment]
> I like it! 

You mean the feature. But also that specific icon? Or should we find a better one?

> I would leave it the way it is. Adding extra checks/conditions seems unnecessary and provides little (if any) benefit. Users are unlikely to trigger the very-long unless intended (even the minimum 2.5s I set it to seems really long to keep the finger in place unintentionally).

Already implemented, was a bit tedious :/ Well, the benefit is just to not bother them with something that make no sense, even if it should happent rarely and unintentionally.
This presence/absence was a bit bothering when testing, as I was waiting expecting it when combinations resulted in it not happening :) but I guess for people who have make their choice of settings, it would be fine to have it mean something consistently.

> By the way, what would you think about offering the whole range of actions for both long press and very-long press? So the user could choose any one for either gesture. 

I would think that it's a bit of too much work for something I won't use :)

> (Personally, I'd rather have the context menu for regular long-press selections and the direct highlight for very-long press.*)

My bet is that you would soon be tired of waiting and end up using short long-press + hit the Highlight button :)

**Comment by jonnyl2** [comment]
>You mean the feature. But also that specific icon? Or should we find a better one?

I meant the icon. That I like the feature I thought was self-evident :)

>This presence/absence was a bit bothering when testing, as I was waiting expecting it when combinations resulted in it not happening :) but I guess for people who have make their choice of settings, it would be fine to have it mean something consistently.

To be honest I would also prefer the icon always showing after very-long-press, even if it wouldn't mean a different action. It's a nice feedback mechanism that shows the system is keeping track of the very-long presses. I can't imagine it bothering someone. I wouldn't even notice it if I didn't force myself to look in the upper-left hand corner. (But I also understand that different people are bothered by very different things.) 

But since you've already done the work it's up to you whether to leave it in or not. 

>>offering the whole range of actions for both long press and very-long press?

>I would think that it's a bit of too much work for something I won't use :)

Fair :)

>My bet is that you would soon be tired of waiting and end up using short long-press + hit the Highlight button :)

I highly doubt that, because there will always be the risk of me missing the Highlight button and then having to redo the whole selection again. But I would hate to give up the single-word dictionary pop-up so it's alright the way it is now :)

By the way, I think it may be a bit more accurate to rename the menu to "Scroll page when gesture position reaches corner" (or something like that). Because the selection could already reach all the way to the corner (e.g., a long last word) but only once you pan into the corner the scroll will trigger. Or the selection may stop in the middle of the last line (if the sentence finishes), but you still have to pan into the corner to trigger the scroll, without the selection going further.

**Comment by poire-z** [comment]
> "Scroll page when gesture position reaches corner"

A bit long, and does not mention "highlight" or "selection". We are indeed in a `Long-press on text >` submenu, but on the 2nd page, and the context may be forgotten. The 2 other menu items mention "highlight", it feels this one should also mention "highlight" or "selection" - and it indeed gets hard to find a small title expressing what this does :/

**Comment by jonnyl2** [comment]
>A bit long, and does not mention "highlight" or "selection"

OK, no problem. It should be self-explanatory.

By the way, I've been wanting to ask about this for a while:

After launching the dictionary from the selection-menu, the selection stays in place even after the dictionary widget is closed. This is intentional, as remarked in the code (L126):

https://github.com/koreader/koreader/blob/5d2a4410648dfc993e6dbc01cbf23941a92404c3/frontend/apps/reader/modules/readerhighlight.lua#L114-L126

However, the reason given is that so that the selection can be highlighted from the dictionary widget (L114-), but this doesn't seem to be possible anyway (Highlight button is disabled)?

**Comment by poire-z** [comment]
"_nearly_ same reason as above" then :)

I think it's because when launching stuff from the text selection buttontable, we don't close this buttontable, as, once the user is done with that launched stuff (ie. Dictionary lookup), he may want to do other stuff (ie. Translate or Wikipedia, or indeed just Highlight, if no Dict lookup result), and it's best to allow that from there than requiring the text selection to be remade. Closing it all is just a tap away.

**Comment by jonnyl2** [comment]
Is there a reason though why the Highlight button doesn't work when the dictionary is called from the buttontable, or when long-press action is set to 'Dictionary'? It works fine on single-word-direct-look ups. (I'm not asking you to fix it; doesn't really matter to me. I'm just curious.)

**Comment by poire-z** [comment]
No real idea, it's been like that for a long time I guess.
Possibly because of the number of items in the chain: text selection > highlight buttontable > dictlookup - makes it complex to handle and to keep a state sane (ie. if highlighted from dict lookup, the buttontable would still show Highlight - or need to be updated to show Unhighight...)

**Comment by jonnyl2** [comment]
I see, thanks.

Btw, a horizontal ellipsis (e.g., Nerdfonts `F141`) might be another idea for the icon in the corner for very-long press. It can somewhat be associated with a context menu, and is also very non-imposing.

<img width="84" alt="image" src="https://github.com/koreader/koreader/assets/95502269/a719657e-d99e-49d2-b158-416ac6c4986e">

The one you picked is fine as well, though.

**Comment by NiLuJe** [comment]
Using something that could arguably be part of the text as an icon could look confusing, I guess?

**Comment by jonnyl2** [comment]
Discovered a (new?) situation in which the poke ball symbol doesn't disappear when it should. Can consistently reproduce on both Kobo and Android:

1. Long-press with one finger until symbol appears
2. Do "Rotation" gesture with a second finger

The symbol only goes away after a new long press gesture on the screen.

**Comment by poire-z** [comment]
I see.
We keep a flag `self.ui.highlight.long_hold_reached`, that ReaderHighlight may unset with `self:_resetHoldTimer()` when some other events that **ReaderHighlight may know** about happen.
But for the other gestures like two finger stuff, ReaderHighlight may not be able to know they are happening (and handled elsewehre).
No real idea (but not much brain bandwidth available) how to solve that. Pinging @NiLuJe in case you have some (idea or bandwidth :)).

**Comment by NiLuJe** [comment]
Yeah, that's probably a bit of a mess. The rotate gesture is responsible for over 50% of the tricky bullshit happening in GestureDetector :/.

Not familiar with the ReaderHighlight parts, but rotate is a *mess* in gestureDetector, so, err, pass? :D.

**Comment by NiLuJe** [comment]
The sanest course of action *might* be making long_hold an actual `GestureDetector` gesture instead of a `ReaderHighlight` hack?

But, yeah, not for the faint of heart ;).

**Comment by jonnyl2** [comment]
Alternatively, just flash the symbol briefly after the time has elapsed, rather than make it sticky until the finger is released?

However, to make it noticeable it should then be closer to the gesture position. On a 6" screen it usually won't matter much, but even on a 7.8" screen the top left corner is too far away when one's eyes are focused on the bottom half of the screen to make a selection (let alone on even bigger devices).
